include/utilities/mdarray/mdarray.hpp

```diff
@@ -0,0 +1,206 @@
+            throw std::out_of_range(
+                "index " + std::to_string(index) +
+                " is larger than maximum possible address index (" +
+                std::to_string(this->max_size()) + ")"
+            );
+        
+        if (index >= this->size() || this->size() == 0)
+            this->m_data.resize(index + 1);
+    
+        this->m_data[index] = value;
+    }
+
+    void insert(std::initializer_list<std::pair<std::initializer_list<size_type>, value_type>> args)
+    {
+        for (const auto& arg : args) {
+            this->insert(arg.first, arg.second);
+        }
+    }
+
+    /**
+     * Allocate this mdarray based on the indices given.
+     * 
+     * @example
+     * mdarray<double> x(3);   // rank 3
+     * x.allocate({5, 10, 10}) // Allocates 5 * 10 * 10 elements
+     * @param n 
+     */
+    void allocate(const boost::span<const size_t> n)
+    {
+        this->m_data.resize(this->index(n));
+    }
+
+    /**
+     * Get the (total) size of this mdarray
+     * (aka the total number of elements).
+     * 
+     * @return size_type
+     */
+    size_type size() const noexcept
+    {
+        return this->m_data.size();
+    }
+
+    size_type max_size() const noexcept
+    {
+        size_type max = 1;
+        for (const auto& v : this->m_shape) {
+            max *= v;
+        }
+        return max;
+    }
+
+    iterator begin() const noexcept {
+        return iterator(*this, 0);
+    }
+
+    iterator end() const noexcept {
+        return iterator(*this, this->m_data.size());
+    }
+
+    /**
+     * Get the rank of this mdarray.
+     * 
+     * @return size_type 
+     */
+    size_type rank() const noexcept
+    {
+        return this->m_shape.size();
+    }
+
+    /**
+     * Get the shape of this mdarray.
+     * 
+     * @return span of the extents per dimension
+     */
+    boost::span<const size_t> shape() const noexcept
+    {
+        return this->m_shape;
+    }
+
+    /**
+     * Index a multi-dimensional set of indices to a single address index.
+     * 
+     * @param n 
+     * @return size_type 
+     */
+    size_type index(const boost::span<const size_t> n) const
+    {
+        size_type index = 0, stride = 1;
+        for (size_type k = 0; k < this->rank(); k++) {
+            index += n[k] * stride;
+            stride *= this->m_shape[k];
+        }
+    
+        return index;
+    }
+
+    /**
+     * Retrieve the multi-dimensional index from a given address index.
+     * 
+     * @param[in] idx Address index (aka flat index)
+     * @param[out] n outputted index
+     */
+    void deindex(size_type idx, boost::span<size_type> n) const
+    {
+        // TODO: Need a bounds check?
+        size_type stride = 1;
+        for (size_type k = 0; k < this->rank(); k++) {
+            n[k] = idx / stride % this->m_shape[k];
+            stride *= this->m_shape[k];
+        }
+    }
+
+  private:
+    std::vector<size_type> m_shape;
+    container_type         m_data;
```

Comment:
I'm a bit lost about the relationship between `m_data.size()` and the values in `m_shape`. The shape is fixed at construction, while `m_data` can grow or shrink arbitrarily with calls to `allocate`. Could you add assertions of the invariants that are actually supposed to hold, and maybe document the enforced relationship in the comments of whichever of `size`, `max_size`, or `allocate` makes sense?

Reply:
An incoming commit handles this by only providing `size` in the public interface. On construction, an `mdarray` will allocate as much space as it needs (via the `std::vector` size constructor) and we'll leave optimizing this to a future concern. The `allocate` member function is removed, and the `max_size` member is converted into a private member inlined at construction.